### ISSUE TITLE
EnumCase.hasAssociatedValue is incorrect for cases with empty parenthesis

### ISSUE THREAD

**Issue by devandsev** [krzysztofzablocki/Sourcery#572]
## Environment
Sourcery 0.10.1
Swift 4.1
Xcode 9.3
macOS 10.13.3

## Problem

Example code:
```swift
enum SomeEnum {
    case a()
    case b
}
```

Swift treats case with `()` as a case with associated value of type `Void`. Sourcery, on the other hand, treats it as a case with no associated values.

Trying to build this code
```swift
let someEnumA = SomeEnum.a()
let someEnumB = SomeEnum.b

if someEnumA == someEnumB {
}
```
produces  Swift error `Binary operator '==' cannot be synthesized for enums with associated values`. Another quick formal proof that there's Void associated value:
```Swift
if case .a(let v) = someEnumA, v is Void {
    print("Yup, there's Void associated type")
}
```

Let's take a look at this Stencil template:
```swift
...
{% if enum.hasAssociatedValues %}
       ...
{% else %}
        return self == anotherCase
{% endif %}
...
```
For `SomeEnum`, Sourcery will generate Swift code which will give us mentioned `Binary operator '==' cannot be synthesized for enums with associated values` error.

## Solution
Add `AssociatedValue` for Void to `EnumCase.associatedValues` when parsing enum cases with empty parenthesis.

**Comment by ilyapuchka** [comment]
Is this something related to changes made in Swift 4.1?

**Comment by devandsev** [comment]
No, reproducing this behaviour in Swift 4.0.3 as well.

**Comment by ilyapuchka** [comment]
Then I guess PR should be based on master branch. 
But I'm not sure if this will not break things. It looks like `case b()` and `case c(Void)` are still different things. First is created with `Foo.b()` and the second is created with `Foo.c(())`, but from both you can extract associated value using pattern matching. So `b()` and `c(Void)` are not exactly the same, so I'm not sure if they should be treated as both having the same associated value. We should have a way to distinguish these cases.

But if we still do it, I think it's better to make type of associated value in this case to be `()` and not `Void`, as you can't do `Foo.c(Void)`, but only `Foo.c(())`.

```swift
enum Foo {
    case a
    case b()
    case c(())
}

let a = Foo.a
let b = Foo.b()
let c = Foo.c(())

if case .a = a {
    print("a has no associated value")
}

if case let .b(v) = b {
    print("b has associated value \(v)")
}

if case let .c(v) = c {
    print("c has associated value \(v)")
}
```

**Comment by devandsev** [comment]
I agree, using `()` is more consistent than `Void`, even though `b` and `c` are still initialized differently. I've updated PR accordingly.

**Comment by ilyapuchka** [comment]
Solved with #573
